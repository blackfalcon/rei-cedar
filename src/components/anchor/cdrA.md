### Modifiers

Modifiers can be combined 1 from each grouping. 

{contrast, overlay, standalone} **OR** button

A modifier list that contains 'button' exposes the same modifiers as cdrButton and the other modifiers shouldn't be combined with it.

### Examples

```
<div>
    <p>Default:</p>
    <cdr-a href="https://www.rei.com/">An example link</cdr-a>

    <p>Contrast:</p>
    <div class="cdr-bg--dark-20">
        <cdr-a modifier="contrast" href="#example-contrast"> An example contrast link</cdr-a>
    </div>

    <p>Overlay:</p>
    <div class="cdr-bg--dark-20">
        <cdr-a modifier="overlay" href="#example-overlay"> An example overlay link</cdr-a>
    </div>

    <p>Standalone:</p>
    <div>
        <cdr-a modifier="standalone" href="https://www.rei.com/">An example standalone link</cdr-a>
    </div>

    <p>Button:</p>
    <cdr-a modifier="button" href="https://www.rei.com/">Link as button</cdr-a>
</div>
```
