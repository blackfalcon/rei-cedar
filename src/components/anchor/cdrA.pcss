.cdr-link {
  cursor: pointer;
  display: inline;
  border: none;
  text-decoration: underline;
  outline: none;
  background-color: $color-transparent-background;
  border-color: $color-transparent-border;
  color: $sys-color-link;

  &:hover,
  &:active,
  &:focus {
    outline: none;
    text-decoration: underline;
    background-color: $color-transparent-background;
    color: $sys-color-link;
  }

  &:focus {
    outline: $outline-default-width $outline-default-style $color-outline-focus;
  }

  &--contrast,
  &--contrast:hover,
  &--contrast:active,
  &--contrast:focus,
  &--contrast:visited {
    color: $sys-color-link-overlay;
  }

  &--overlay,
  &--overlay:hover,
  &--overlay:active,
  &--overlay:focus,
  &--overlay:visited {
    color: $sys-color-link-contrast-overlay;
  }

  &--standalone {
    text-decoration: none;

    &:hover,
    &:active,
    &:focus,
    &:visited {
      text-decoration: underline;
    }
  }
}
