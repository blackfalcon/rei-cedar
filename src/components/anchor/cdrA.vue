<template>
  <a :class="modifierClass" :target="target" :rel="newRel">
    <slot></slot>
  </a>
</template>

<script>
  import modifier from '../../mixins/modifier';
  
  export default {
    name: 'cdr-a',
    mixins: [modifier],
    props: {
      target: String,
      rel: String,
    },
    computed: {
      baseClass() {
        const modifiers = this.modifier ? this.modifier.split(' ') : [];
        return modifiers.indexOf('button') >= 0 ? 'cdr-button' : 'cdr-link';
      },
      newRel() {
        if (this.target === '_blank') {
          return this.rel || 'noopener noreferrer';
        }
        return this.rel;
      },
    },
  };
</script>
