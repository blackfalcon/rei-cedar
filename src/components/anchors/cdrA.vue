<template>
  <a :class="modifierClass" :target="target" :rel="rel">
    <slot></slot>
  </a>
</template>

<script>
  import modifier from '../../mixins/modifier';
  
  export default {
    name: 'cdrA',
    mixins: [modifier],
    props: {
      target: String,
      rel: String,
    },
    computed: {
      baseClass() {
        const modifiers = this.modifier ? this.modifier.split(' ') : [];
        return modifiers.indexOf('button') >= 0 ? 'cdr-button' : 'cdr-link';
      },
      newRel() {
        if (this.target === '_blank') {
          return this.rel || 'noopener';
        }
        return this.rel;
      },
    },
  };
</script>
