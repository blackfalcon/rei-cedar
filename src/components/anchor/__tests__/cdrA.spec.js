import { shallow } from 'vue-test-utils';
import { createRenderer } from 'vue-server-renderer'
import cdrA from 'Components/anchor/cdrA';

describe('cdrA.vue', () => {
  it('outputs the same HTML', () => {
    const wrapper = shallow(cdrA);
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    });
  });

  it('renders an anchor', () => {
    const wrapper = shallow(cdrA);
    expect(wrapper.is('a')).toBe(true);
  });

  it('sets target attr correctly', () => {
    const wrapper = shallow(cdrA, {
      propsData: {
        target: '_self',
      },
    });
    expect(wrapper.hasAttribute('target', '_self')).toBe(true);
  });

  it('sets rel attr correctly', () => {
    const wrapper = shallow(cdrA, {
      propsData: {
        rel: 'nofollow',
      },
    });
    expect(wrapper.hasAttribute('rel', 'nofollow')).toBe(true);
  });

  it('computes target="_blank" rel attr correctly', () => {
    const wrapper = shallow(cdrA, {
      propsData: {
        target: '_blank',
      },
    });
    expect(wrapper.hasAttribute('rel', 'noopener noreferrer')).toBe(true);
  });

  it('computes base class correctly', () => {
    const wrapper = shallow(cdrA, {
      propsData: {
        modifier: 'primary',
      },
    });
    expect(wrapper.hasClass('cdr-link')).toBe(true);
  });

  it('computes link as button base class correctly', () => {
    const wrapper = shallow(cdrA, {
      propsData: {
        modifier: 'primary, button',
      },
    });
    expect(wrapper.hasClass('cdr-button')).toBe(true);
  });
});