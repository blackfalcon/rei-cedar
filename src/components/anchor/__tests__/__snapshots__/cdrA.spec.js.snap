// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdrA.vue outputs the same HTML 1`] = `<a class="cdr-link"></a>`;
